dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/relationship/RelationshipDeletionHandler.java

```diff
@@ -76,6 +78,37 @@ public void deleteTrackedEntityInstance( TrackedEntityInstance entityInstance )
             }
         }
     }
+    
+    @Override
+    public void deleteProgramStageInstance( ProgramStageInstance programStageInstance )
+    {
+        Collection<Relationship> relationships = relationshipService
+            .getRelationshipsByProgramStageInstance( programStageInstance, false );
+
+        if ( relationships != null )
+        {
+            for ( Relationship relationship : relationships )
+            {
+                relationshipService.deleteRelationship( relationship );
+            }
+        }
```

Comment:
just as a suggestion:
we might replace this block of code which is duplicated across the whole class with a private method like:
``` java
    private void deleteRelationships(Collection<Relationship> relationships )
    {
        Optional.ofNullable( relationships )
            .orElse( Collections.emptySet() )
            .forEach( relationshipService::deleteRelationship );
    }
```
so that each method would become just:
``` java
    @Override
    public void deleteProgramStageInstance( ProgramStageInstance programStageInstance )
    {
        deleteRelationships( relationshipService.getRelationshipsByProgramStageInstance( programStageInstance, false ) );
    }
```